### PR TITLE
Patch with more than10 column

### PR DESCRIPTION
**Pull request by xinlian12**
**Issue:**

- With the current ItemPatch strategy, customer can only patch up to 10 columns (https://learn.microsoft.com/en-us/azure/cosmos-db/partial-document-update#supported-modes). However, there are cases where customer need to patch more than 10 columns. 

**Solution**
`ItemBulkUpdate` is being introduced as a new strategy. 
Internally, SDK gonna utilize readMany to read items back, update items and then write item back to CosmosDb.

- If the item existed in the Cosmos DB already, then it will be updated based on etag precondition
- If the item does not exist in the CosmosDB, then it will be created with all updated properties 

**Test**
Using the NY taxi data sample, ingest 10 Million records by using the new `ItemBatchBulkUpdate` strategy. Which took around 27 minute to finish. (using itemOverwritten strategy took 25 mins to finish)
image

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala [hunk 1]
```diff
@@ -432,29 +654,45 @@ class BulkWriter(container: CosmosAsyncContainer,
...
+    val operationsLog = getActiveOperationsLog(activeOperationsSnapshot, activeReadManyOperationsSnapshot)
 
-    if (activeOperationsSnapshot.equals(activeOperations)) {
+    if (activeOperationsSnapshot.equals(activeBulkWriteOperations)
+        && activeReadManyOperationsSnapshot.equals(activeReadManyOperations)) {
       numberOfIntervalsWithIdenticalActiveOperationSnapshots.incrementAndGet()
       log.logWarning(
         s"${operationName} has been waiting ${numberOfIntervalsWithIdenticalActiveOperationSnapshots} " +
-          s"times for identical set of operations: ${operationsLog} " +
+          s"times for identical set of bulkWrite operations: ${operationsLog} " +
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:689]: revert back the message

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala]: Done inline

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala [hunk 2]
```diff
@@ -130,6 +143,202 @@ class BulkWriter(container: CosmosAsyncContainer,
...
+                                  .get
+                                  .createCosmosPatchBulkUpdateOperations(readManyOperation.objectNode)
+
+                          val rootNode =
+                              cosmosPatchHelperOpt
+                                  .get
+                                  .patchBulkUpdateItem(resultMap.get(readManyOperation.cosmosItemIdentity), cosmosPatchBulkUpdateOperations)
+
+                          // create bulk item operation
+                          val etagOpt = Option.apply(rootNode.get(CosmosConstants.Properties.ETag))
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:211]: Question: Is this done to match Spark 2 - would we allow to configure whether to use etag precondition always (I see why it makes sense - so no criticism - just trying to fully understand the intend here)

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala [hunk 3]
```diff
@@ -130,6 +143,202 @@ class BulkWriter(container: CosmosAsyncContainer,
...
+                                      readManyOperation.operationContext.attemptNumber,
+                                      Some(readManyOperation.objectNode)
+                                  ))
+                          }
+
+                          this.emitBulkInput(bulkItemOperation)
+                      }
+                  })
+                  .onErrorResume(throwable => {
+                      for (readManyOperation <- readManyOperations) {
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:242]: QUESTION: This error handling seems fine - but wouldn't we also need ot make sure that for certain errors 409 on create and 412 on Replace the retry would redo the read-many as well?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:242]: What i mean is I saw that you added the error handling on patch in shoudlRetry for 409/412 - but the errors wouldn't happen / be caught here - because you just schedule the bulk operation?

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala [hunk 4]
```diff
@@ -872,7 +874,8 @@ private object CosmosWriteConfig {
       "ignore pre-existing items i.e., Conflicts), `ItemDelete` (deletes based on id/pk of data frame), " +
       "`ItemDeleteIfNotModified` (deletes based on id/pk of data frame if etag hasn't changed since collecting " +
       "id/pk), `ItemOverwriteIfNotModified` (using create if etag is empty, update/replace with etag pre-condition " +
-      "otherwise, if document was updated the pre-condition failure is ignored)")
+      "otherwise, if document was updated the pre-condition failure is ignored)," +
+      " `ItemPatchBulkUpdate` (read item, then patch the item locally, then using create if etag is empty, update/replace with etag pre-condition)")
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:878]: I think it would be good to elaborate a bit on automatic retries - I would assume if a competing update between read many and create or replace happens we would automatically retry and start over from read many - so, this is informative for customers to understand how we achieve that we can safely patch the document. But nothing they should need to worry about.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala (hunk 3 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:242]: yea we will not get 409 or 412 in readMany, it will only happen during bulk write.

in the shouldRetry logic, it added the 409 and 412 checking for patchBulkUpdate.  so when an operation failed during the bulkWrite step and when we retry, it will go through the whole flow again -> first readmany then bulkWrite

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala [hunk 5]
```diff
@@ -162,4 +165,113 @@ class CosmosPatchHelper(diagnosticsConfig: DiagnosticsConfig,
...
+                  case _ => throw new RuntimeException(s"Unsupported parent node type ${parentNode.getClass} in patchBulkUpdateFields") // we should never reach here
+              }
+
+              rootNode
+          case _ => throw new RuntimeException(s"PatchOperationType $patchOperationType is not supported for ItemPatchBulkUpdate") // we should have never reach here
+      }
+  }
+
+  private[this] def getOrCreateNextParentNode(parentNode: JsonNode, childPath: String): JsonNode = {
+      val jsonPath = JsonPointer.compile(childPath)
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala:256]: QUESTION: compile sounds awfully expensive (worth caching?) - but have never used JsonPointer data type.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala [hunk 6]
```diff
@@ -162,4 +165,113 @@ class CosmosPatchHelper(diagnosticsConfig: DiagnosticsConfig,
...
+      }
+
+      patchBulkUpdateOperations.foreach(patchOperation => {
+          patchBulkUpdateFields(rootNode, patchOperation.getOperationType, patchOperation.getPath, patchOperation.getResource)
+      })
+
+      rootNode
+  }
+
+  private[this] def patchBulkUpdateFields(
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala:223]: NIT - this method and the one below look like it would be good to have explicit unit testing for them. I would consider making them package private and add explicit unit tests just for each of these two methods

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala [hunk 7]
```diff
@@ -332,6 +337,62 @@ class PointWriter(container: CosmosAsyncContainer,
...
+              val updatedNode = cosmosPatchHelpOpt.get.patchBulkUpdateItem(Some(response.getItem), patchBulkUpdateOperations)
+              val cosmosItemRequestOptions = new CosmosItemRequestOptions().setIfMatchETag(response.getETag)
+              container.replaceItem(updatedNode, itemId, partitionKey, cosmosItemRequestOptions)
+          })
+          .asScala
+          .onErrorResume(throwable => {
+              throwable match {
+                  case e: CosmosException if Exceptions.isNotFoundException(e) =>
+                      val updatedNode = cosmosPatchHelpOpt.get.patchBulkUpdateItem(None, patchBulkUpdateOperations)
+                      container.createItem(updatedNode, partitionKey, new CosmosItemRequestOptions()).asScala
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala:389]: This one can now result in 409 which would need to be retried again

**Review by FabianMeiswinkel — COMMENTED** [review]
Looks great - no blocking comments - just few questions that we should  probably discuss in a call before merging.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:211]: Never mind - etag definitely is needed - you just need to make sure the 409/412 handling includes the read-many on retry

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:242]: Got it - yes, should work as is

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala (hunk 4 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala]: updated

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala (hunk 6 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala:223]: still keep this method object private level, but added tests for `patchBulkUpdateItem` which will internally call these two methods

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala (hunk 7 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala:391]: good catch, added checking for 409/412 as well

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala [hunk 8]
```diff
@@ -130,6 +143,219 @@ class BulkWriter(container: CosmosAsyncContainer,
...
+      // If in the future, there is a need to separate the configuration, can re-consider
+      val bulkBatchSize = ImplementationBridgeHelpers
+          .CosmosBulkExecutionOptionsHelper
+          .getCosmosBulkExecutionOptionsAccessor
+          .getMaxMicroBatchSize(cosmosBulkExecutionOptions)
+      val batchInterval = ImplementationBridgeHelpers
+          .CosmosBulkExecutionOptionsHelper
+          .getCosmosBulkExecutionOptionsAccessor
+          .getMaxMicroBatchInterval(cosmosBulkExecutionOptions)
+      val batchConcurrency = ImplementationBridgeHelpers
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:166]: limiting the concurrency as well

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala (hunk 5 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala:256]: the compile method seems pretty lightweight, so will keep it for now

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java [hunk 9]
```diff
@@ -1440,9 +1440,46 @@ public <T> Mono<FeedResponse<T>> readMany(
...
+     *     .subscribe&#40;&#41;;
+     * </pre>
+     * <!-- end com.azure.cosmos.CosmosAsyncContainer.readMany -->
+     * @param <T> the type parameter
+     * @param itemIdentityList CosmosItem id and partition key tuple of items that that needs to be read
+     * @param requestOptions the optional request option
+     * @param classType   class type
+     * @return a Mono with feed response of cosmos items or error
+     */
+    <T> Mono<FeedResponse<T>> readMany(
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:1473]: need to add this so to use throughput control

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java [hunk 10]
```diff
@@ -2589,14 +2588,6 @@ private SqlQuerySpec createReadManyQuerySpecPartitionKeyIdSame(
             String idValue = itemIdentity.getId();
             String idParamName = "@param" + i;
 
-            PartitionKey pkValueAsPartitionKey = itemIdentity.getPartitionKey();
-            Object pkValue = ModelBridgeInternal.getPartitionKeyObject(pkValueAsPartitionKey);
-
-            if (!Objects.equals(idValue, pkValue)) {
-                // this is sanity check to ensure id and pk are the same
-                continue;
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java:2597]: the partitionKey created from spark side does not populate objectValue.
In order to let readmany work properly, either we change from spark side, or remove the checking here. 

but instead of SDK silently suppress this issue, it might be better to let backend throw exception if the data input from customer is invalid? Open for suggestions

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala [hunk 11]
```diff
@@ -130,6 +143,219 @@ class BulkWriter(container: CosmosAsyncContainer,
...
+                                              readManyOperation.operationContext.attemptNumber,
+                                              Some(readManyOperation.objectNode)
+                                          ))
+                                  case None => CosmosBulkOperations.getCreateItemOperation(
+                                      rootNode,
+                                      readManyOperation.cosmosItemIdentity.getPartitionKey,
+                                      OperationContext(
+                                          readManyOperation.operationContext.itemId,
+                                          readManyOperation.operationContext.partitionKeyValue,
+                                          None,
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:248]: NIT: variable name like in c# is possible in scala and would help readability here

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala [hunk 12]
```diff
@@ -86,6 +86,7 @@ private[spark] object CosmosConfigNames {
   val WritePatchDefaultOperationType = "spark.cosmos.write.patch.defaultOperationType"
   val WritePatchColumnConfigs = "spark.cosmos.write.patch.columnConfigs"
   val WritePatchFilterPredicate = "spark.cosmos.write.patch.filter"
+  val WritePatchBulkUpdateColumnConfigs = "spark.cosmos.write.patchBulkUpdate.columnConfigs"
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:89]: Please remove the patch prefix here - we should call the feature publicly "BulkUpdate" - not patchBulUpdate - otherwise the whole discussion why not all patch operations are possibel etc. will come up

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala (hunk 5 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala:256]: Please add a comment - saves the next person to investigate again :-)

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala [hunk 13]
```diff
@@ -332,6 +336,65 @@ class PointWriter(container: CosmosAsyncContainer,
...
+  // scalastyle:on multiple.string.literals
+  private def patchBulkUpdateWithRetry(
+                                      partitionKeyValue: PartitionKey,
+                                      objectNode: ObjectNode): Unit = {
+
+      var exceptionOpt = Option.empty[Exception]
+      val patchBulkUpdateOperation = PatchBulkUpdateOperation(taskDiagnosticsContext,
+          CosmosItemIdentifier(objectNode.get(CosmosConstants.Properties.Id).asText(), partitionKeyValue))
+
+      for (attempt <- 1 to cosmosWriteConfig.maxRetryCount + 1) {
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala:348]: why not 0 to maxRetryCount and save the whole + 1 magic?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java (hunk 10 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentClientImpl.java:2597]: Agreed

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala (hunk 5 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala:260]: updated

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala (hunk 12 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala]: make sense, updated the public area to ItemBulkUpdate (The internal methods still prefixed with patch)

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala [hunk 14]
```diff
@@ -3,15 +3,18 @@
...
-import com.azure.cosmos.implementation.{Constants, ImplementationBridgeHelpers, Utils}
 import com.azure.cosmos.implementation.apachecommons.lang.StringUtils
+import com.azure.cosmos.implementation.patch.{PatchOperationCore, PatchOperationType}
+import com.azure.cosmos.implementation.{Constants, ImplementationBridgeHelpers, Utils}
 import com.azure.cosmos.models.{CosmosPatchOperations, PartitionKeyDefinition}
 import com.azure.cosmos.spark.CosmosPredicates.{assertNotNull, assertNotNullOrEmpty}
 import com.azure.cosmos.spark.diagnostics.LoggerHelper
-import com.fasterxml.jackson.databind.{JsonNode, ObjectMapper}
+import com.fasterxml.jackson.core.JsonPointer
 import com.fasterxml.jackson.databind.node._
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala:13]: NIT NIT: Avoid the wildcard import (only even look at it when you make another commit anyway :-))

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala (hunk 13 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala:348]: But even there without the + 1 magic it would perfectly ok, 

for (attempt <- 1 to cosmosWriteConfig.maxRetryCount + 1) {
log.logWarning( s"patch update item $patchBulkUpdateOperation attempt #$attempt max remaining retries " + s"${cosmosWriteConfig.maxRetryCount + attempt}, encountered ${e.getMessage}")

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala:348]: Only diff would be that the pound=prefixed index starts 0 absed?

Also a NIT by the way

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - thanks!

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala (hunk 14 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPatchHelper.scala]: updated

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala (hunk 13 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PointWriter.scala]: updated. the same pattern also used for other retry methods, so updated there as well

**Comment by xinlian12** [comment]
Tested locally for the failed tests:
CosmosItemWriteRetriesTest::deleteItem
For the failed spark 3.4 tests -> it is because there is no cluster setup yet, will follow up in a different PR
